text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+to the intermediary.
+
+#### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new messages.
+
+The consumer receives, processes, and settles a message.
+
+* "Receiving" is the process of obtaining a message from the intermediary.
+* "Processing" is the process of acting on the information a message contains.
+* "Settling" is the process where intermediary and consumer agree on the state
+  of the transfer.
+
+#### Intermediary
+
+An "intermediary" receives a message for the purpose of forwarding it to the
+next receiver, which might be another intermediary or a consumer.
+
+### Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```
+                               CONSUMER
+
+                   . . . . . . Settle
+PRODUCER           .              ^
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extracted from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+> A producer SHOULD attach a creation context to each message. The creation context
+> SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
```

Comment:
As stated, the creation context is intended to allow correlation of a message's producer with its consumers. One of the criticisms of the concept of the creation span was that it is essentially a zero-duration span, that isn't really measuring anything. The counter to that criticism was that producer-consumer correlation is important and such a context is important for creating these links.

This makes me wonder why a message's send span couldn't serve this purpose? I believe the main concern with this was that a send span may represent a batch of messages. It should be possible to precisely correlate back to a unique context for a single message.

During yesterday's discussion, we were further puzzling over batch send spans and how to express per-message attributes such as message_id, correlation_id, and destination.name. A number of possibilities were discussed, but a couple that I'll highlight were:
1. Do not include per-message attributes in the send span directly. Instead, for each message, create a link to its creation context. Inside the link, include per-message attributes.
2. Create a span for each message in the batch as a child of the batch send. For some messaging systems, this may be a zero-duration span again. However, for other systems I believe a batch send may simply be a convenience function that loops over sending individual messages. In these cases the spans could measure sending the individual messages. Perhaps they are acknowledged as a batch and therefore they would all end at the same time. But you could measure from when you started the send of that message until the batch was acked.

I believe we generally agreed option 2 was the nicest option. The concern was the cost of extra spans and the general dislike for zero-duration spans. However, what if this span also served as the creation context? This gets rid of a span that is almost certainly zero-duration with a span that:
* Measures an actual send operation for single message sends
* May measure an actual send operation for a single message in a batch; or
* May be a zero-duration span for other messages. In these cases wouldn't it also be an option to have the span duration cover the batch send? I'm not sure it adds a lot of value, but I also don't see a problem with this approach.

I think if we took this approach, it completely gets rid of all zero-duration spans for single message sends. It may get rid of zero duration spans for batch sends. I think in all cases the total number of spans is the same or less than we would have had otherwise, and the overal span structure modeling makes a lot more sense to me.

If we did this, I think it makes sense for the "batch send" span (which is the parent of all individual message send spans) be defined as it's own operation, and with its own set of attributes. It would likely be lighter weight as it would only include attributes common for all messages in the send. I think this span structure would be easier for back ends to implement against.

Is there anything I'm missing as to why a send span couldn't serve as the context used for producer-consumer correlation?

Does anyone have any feedback on this idea?

Reply:
@dpauls thanks for putting the summary here!

> However, for other systems I believe a batch send may simply be a convenience function that loops over sending individual messages.

This is interesting. In this case the span/context in each message is actually something that can be used to track time. Do you happen to know maybe a system/SDK that does this?

> I believe we generally agreed option 2 was the nicest option. The concern was the cost of extra spans and the general dislike for zero-duration spans. However, what if this span also served as the creation context? This gets rid of a span that is almost certainly zero-duration with a span that:


As we discussed, my understanding is that a creation context _is_ a span, created for each message as part of a batch send operation (thus being children of the "send" span). It's also what the Johannes wrote in this OTEP here: https://github.com/pyohannes/oteps/blob/72e1215d698ad86473f874ef0f0282de07c66521/text/trace/0192-messaging-semantic-conventions-spec.md?plain=1#L152

What I'm a bit confused now is this part: https://github.com/pyohannes/oteps/blob/72e1215d698ad86473f874ef0f0282de07c66521/text/trace/0192-messaging-semantic-conventions-spec.md?plain=1#L162

> If a "Create" span exists for a message, its context SHOULD be attached to
> the message. If no "Create" span exists, the context of the related "Publish"
> span SHOULD be attached to the message.

I interpret this as:

- Single message: The `Publish` span is attached to the message, essentially being copied to it. This makes sense, as it's not necessary to have an extra create span for this case. We can attach message-specific attributes here.

- Batch: If the messages don't have the create context, then the `Publish` span context should be attached to the messages in the batch. Then here we run into the message-specific attributes problem.

Reply:
I want to mention that a create-span is not necessarily zero-duration. I think there is an argument for changing the concept of creation to make the stages more symmetrical so that just how "publish" corresponds to "receive" or "deliver", "create" should correspond to "process". Meaning, "create" should track the logical, application-level business logic of creating the message (data). The following would then happen in order:

1. (n times) create some logical message, store creation context in there
2. publish one of these previously created messages, or a whole batch. Notice that this will typically not be a child of the create span, but might even be on a separate trace. If a creation context is not already on the message, you would still have to proceed as currently discussed, but if there is, no artificial spans have to be created.
3. ... from here one as before ...

Just like for a "process" span, a natural "create" span will usually require some cooperation by the application developer and in many cases cannot be automatically created.

Reply:
I guess we all on the same page that having unqiue context per message is the best approach here. And setting atrificial duration on such spans would be a way to hide logical nature of this span - in most cases it does not track any work or operation except message creation. The only scenario where it'd provide value is when messages are sent separately by the client SDK and it seems like an edge case.

I suggest we don't try to hide this fact - we're lacking OTel concept such as unique context without a span or span without duration or event with unique context and we should discuss it with the broader  OTel spec community before making any decisions here.

Reply:
Summarized different options here https://docs.google.com/document/d/1OrHsepd6GjzXKll1ggZyx1jBQd0d_t8NZXT1ZOem7D0/edit?usp=sharing - let's see if we can discuss it tomorrow on the call.

Reply:
>  Meaning, "create" should track the logical, application-level business logic of creating the message (data)

@Oberon00 's take is indeed interesting,  I think we never thought this way. It highly depends on how the application works but I can imagine something like this:

- App receives a request to schedule a birthday notification to a user
- App needs to find the user birth date to be able to create the message `Start create context here`
- App fetchs the necessary data
- App creates the message, with the user data + adds the ambient/current create span to it
- App passes the message on to be send
- Scheduler service receives the message and process it

This indeed is not a 0 duration span, but of course highly depends on application owners to create such spans. In other cases, creating a message might be really just `var myMessage = new Message(data)` so it's really a `0` duration.

Reply:
The question is then if creation of `data` is already (part of) the logical message creation.

Reply:
> The question is then if creation of data is already (part of) the logical message creation.

this is very app-specific. Should message span be the span in which data was read? or created? Both seem to be controversial.

Auto-instrumentations can only control what happens within SDK API and at least my hopes of manual instrumentation precisely following conventions are low.

Reply:
From my understanding, we are going to push for a message span that goes from producer send to consumer receive. Right?